File: Blind 75/16 - Intervals/56-merge-intervals.py
'''
Given an array of intervals where intervals[i] = [start_i, end_i], merge all overlapping intervals, and return an array of the non-overlapping intervals that cover all the intervals in the input.
'''
from time import time
from typing import List


class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        res = []
        for interval in sorted(intervals, key=lambda i: i[0]):
            if res and interval[0] <= res[-1][1]:
                res[-1][1] = max(res[-1][1], interval[1])
            else:
                res.append(interval)
        return res

    def reference(self, intervals: List[List[int]]) -> List[List[int]]:
        intervals.sort(key=lambda pair: pair[0])
        output = [intervals[0]]

        for start, end in intervals:
            lastEnd = output[-1][1]

            if start <= lastEnd:
                # Merge
                output[-1][1] = max(lastEnd, end)
            else:
                output.append([start, end])
        return output

    def quantify(self, test_cases, runs=50000):
        sol_start = time()
        for i in range(runs):
            for case in test_cases:
                if i == 0:
                    print(self.merge(case))
                else:
                    self.merge(case)
        print(f'Runtime for our solution: {time() - sol_start}')

        ref_start = time()
        for i in range(0, runs):
            for case in test_cases:
                if i == 0:
                    print(self.reference(case))
                else:
                    self.reference(case)
        print(f'Runtime for reference: {time() - ref_start}')


if __name__ == '__main__':
    test = Solution()
    test_cases = [
        [[1, 3], [2, 6], [8, 10], [15, 18]],
        [[1, 4], [4, 5]],
        # Additional
        [[1, 4], [0, 4]]
    ]
    test.quantify(test_cases)

File: Blind 75/16 - Intervals/JavaScript/56-merge-intervals.js
const { performance } = require("perf_hooks");

class Solution {
  method(intervals) {
    intervals.sort((a, b) => a[0] - b[0]);
    let res = [intervals[0]];

    for (let i = 1; i < intervals.length; i++) {
      const lastInterval = res[res.length - 1];
      if (lastInterval[1] >= intervals[i][0]) {
        lastInterval[1] = Math.max(lastInterval[1], intervals[i][1]);
      } else {
        res.push(intervals[i]);
      }
    }

    return res;
  }

  reference(intervals) {
    intervals.sort(([aStart, aEnd], [bStart, bEnd]) =>
      aStart !== bStart ? aStart - bStart : aEnd - bEnd
    );

    return this.mergerInterval(intervals);
  }

  mergerInterval = (intervals, merged = []) => {
    let prev = intervals.shift();

    for (const curr of intervals) {
      const [prevStart, prevEnd] = prev;
      const [currStart, currEnd] = curr;

      const hasOverlap = currStart <= prevEnd;
      if (hasOverlap) {
        prev[1] = Math.max(prev[1], curr[1]);
        continue;
      }

      merged.push(prev);
      prev = curr;
    }

    return [...merged, prev];
  };

  quantify(testCases, runs = 1e6) {
    const runsArr = Array.from({ length: runs });
    const solStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.method(input));
        else this.method(input);
      });
    });
    console.log(
      `Runtime for solution: ${(performance.now() - solStart) / 1000}\n`
    );

    const refStart = performance.now();
    runsArr.map((_, i) => {
      testCases.map((input) => {
        if (i === 0) console.log(this.reference(input));
        else this.reference(input);
      });
    });
    console.log(
      `Runtime for reference: ${(performance.now() - refStart) / 1000}`
    );
  }
}

const test = new Solution();
const testCases = [
  [
    [1, 3],
    [2, 6],
    [8, 10],
    [15, 18],
  ],
  [
    [1, 4],
    [4, 5],
  ],
  // Additional
  [
    [1, 4],
    [0, 4],
  ],
];
test.quantify(testCases);
